File: src/main/java/coyote/dataframe/marshal/JSONMarshaler.java
/*
 * Copyright (c) 2014 Stephan D. Cote' - All rights reserved.
 * 
 * This program and the accompanying materials are made available under the 
 * terms of the MIT License which accompanies this distribution, and is 
 * available at http://creativecommons.org/licenses/MIT/
 *
 * Contributors:
 *   Stephan D. Cote 
 *      - Initial concept and initial implementation
 */
package coyote.dataframe.marshal;

import java.io.BufferedWriter;
import java.io.IOException;
import java.io.StringWriter;
import java.util.List;

import coyote.dataframe.DataField;
import coyote.dataframe.DataFrame;
import coyote.dataframe.marshal.json.JsonFrameParser;
import coyote.dataframe.marshal.json.JsonWriter;
import coyote.dataframe.marshal.json.WriterConfig;


/**
 * 
 */
public class JSONMarshaler {

  /**
   * Marshal the given JSON into a dataframe.
   * 
   * @param json
   * 
   * @return Data frame containing the JSON represented data
   */
  public static List<DataFrame> marshal( final String json ) throws MarshalException {
    List<DataFrame> retval = null;

    try {
      retval = new JsonFrameParser( json ).parse();
    } catch ( final Exception e ) {
      System.out.println( "oops: " + e.getMessage() );
      throw new MarshalException( "Could not marshal JSON to DataFrame", e );
    }

    return retval;
  }




  /**
   * Generate a JSON string from the given data frame.
   * 
   * @param frame The frame to marshal
   * 
   * @return A JSON formatted string which can be marshaled back into a frame
   */
  public static String marshal( final DataFrame frame ) {
    return write( frame, WriterConfig.MINIMAL );
  }




  /**
   * Generate a nicely formatted (and indented) JSON string from the given data frame.
   * 
   * @param frame The frame to marshal
   * 
   * @return A JSON formatted string which can be marshaled back into a frame
   */
  public static String toFormattedString( final DataFrame frame ) {
    return write( frame, WriterConfig.PRETTY_PRINT );
  }




  /**
   * @param frame
   * @param config
   * @return
   */
  private static String write( final DataFrame frame, final WriterConfig config ) {

    // create string writer
    final StringWriter sw = new StringWriter();
    final BufferedWriter bw = new BufferedWriter( sw );
    final JsonWriter writer = config.createWriter( bw );

    try {
      writeFrame( frame, writer );
      bw.flush();
    } catch ( IOException e ) {
      return "[\"" + e.getMessage() + "\"]";
    }
    return sw.getBuffer().toString();
  }




  /**
   * 
   * @param frame
   * @param config
   * 
   * @return
   * @throws IOException 
   */
  private static void writeFrame( final DataFrame frame, final JsonWriter writer ) throws IOException {

    if ( frame.size() > 0 ) {
      boolean isArray = frame.isArray();
      if ( isArray )
        writer.writeArrayOpen();
      else
        writer.writeObjectOpen();

      DataField field = null;
      for ( int i = 0; i < frame.size(); i++ ) {
        field = frame.getField( i );

        if ( !isArray ) {
          if ( field.getName() != null ) {
            writer.writeString( field.getName() );
          } else {
            writer.writeString( "" );
          }
          writer.writeMemberSeparator();
        }

        if ( field.getType() == DataField.NULLTYPE ) {
          writer.writeLiteral( "null" );
        } else if ( field.getType() == DataField.BOOLEANTYPE ) {
          if ( "true".equalsIgnoreCase( field.getStringValue() ) ) {
            writer.writeLiteral( "true" );
          } else {
            writer.writeLiteral( "false" );
          }
        } else if ( field.isNumeric() ) {
          writer.writeNumber( field.getStringValue() );
        } else if ( field.getType() == DataField.FRAMETYPE ) {
          writeFrame( (DataFrame)field.getObjectValue(), writer );
        } else {

          writer.writeString( field.getObjectValue().toString() );
        }
        if ( i + 1 < frame.size() ) {
          writer.writeObjectSeparator();
        }
      }

      if ( isArray )
        writer.writeArrayClose();
      else
        writer.writeObjectClose();

    } else {
      writer.writeObjectOpen();
      writer.writeObjectClose();
    }

    return;
  }
}

File: csharp/DataFrame/Coyote/DataFrame/JSONMarshaler.cs
#region license
// Copyright (c) 2006 Stephan D. Cote' - All rights reserved.
// 
// This program and the accompanying materials are made available under the 
// terms of the MIT License which accompanies this distribution, and is 
// available at http://creativecommons.org/licenses/MIT/
// *
// Contributors:
//   Stephan D. Cote 
//      - Initial API and implementation
#endregion
using System;
using System.IO;
using System.Collections.Generic;
using Coyote.DataFrame.JSON;

namespace Coyote.DataFrame {

    ///<summary>
    /// Marshal JSON into and out of a DataFrame.
    ///</summary> 
    public class JSONMarshaler {
        private const string NULL = "null";
        private const string TRUE = "true";
        private const string FALSE = "false";



        /// <summary>
        /// Marshal the given JSON into a dataframe.
        /// </summary>
        /// <param name="json">the JSON string to marshal</param>
        /// <returns>A list of dataframes representing all the objects in the string</returns>
        /// <exception cref="MarshalException">If there was a problem parsing the JSON string.</exception>
        public static List<DataFrame> Marshal( string json ) {
            List<DataFrame> retval = null;

            try {
                retval = new JsonFrameParser( json ).Parse();
            } catch ( Exception e ) {
                throw new MarshalException( "Could not marshal JSON to DataFrame: " + e.Message, e );
            }

            return retval;
        }



        
        /// <summary>
        /// Generate a JSON string from the given data frame.
        /// </summary>
        /// <param name="frame">The frame to marshal</param>
        /// <returns>A JSON formatted string which can be marshaled back into a frame</returns>
        public static string Marshal( DataFrame frame ) {
            return write( frame, JsonWriterConfig.MINIMAL );
        }



        
        /// <summary>
        /// Generate a nicely formatted (and indented) JSON string from the given data frame.
        /// </summary>
        /// <param name="frame">The frame to marshal</param>
        /// <returns>A JSON formatted string which can be marshaled back into a frame</returns>
        public static string ToFormattedString( DataFrame frame ) {
            return write( frame, JsonWriterConfig.FORMATTED );
        }




        /// </summary>
        /// Write the given frame using the given JSON writer.
        /// </summary>
        /// <param name="frame">The DataFrame to be written</param>
        /// <param name="config">The configuration of a JSON writer to use when writing the data</param>
        /// <returns>A JSON string representing the results of writing the data. It may contain error text if an I/O exception occurred</returns>
        private static string write( DataFrame frame, JsonWriterConfig config ) {

            // create string writer
            StringWriter sw = new StringWriter();
            JsonWriter writer = config.CreateWriter( sw );

            try {
                writeFrame( frame, writer );
                sw.Flush();
            } catch ( IOException e ) {
                return "[\"" + e.Message + "\"]";
            }
            return sw.ToString();
        }




        /// <summary>
        /// Write the given frame using the given JSON writer. This is expected to be called recursively.
        /// </summary>
        /// <param name="frame">The DataFrame to be written</param>
        /// <param name="writer">The instance of the JSON writer responsible for writing the data</param>
        private static void writeFrame( DataFrame frame, JsonWriter writer ) {

            if ( frame == null || writer == null ) {
                return;
            }

            if ( frame.Size > 0 ) {
                bool isArray = frame.IsArray;
                if ( isArray )
                    writer.writeArrayOpen();
                else
                    writer.writeObjectOpen();

                DataField field = null;
                for ( int i = 0; i < frame.Size; i++ ) {
                    field = frame.Field[i];

                    if ( !isArray ) {
                        if ( field.Name != null ) {
                            writer.writeString( field.Name );
                        } else {
                            writer.writeString( "" );
                        }
                        writer.writeMemberSeparator();
                    }

                    if ( field.Type == DataField.UDEF ) {
                        writer.writeLiteral( NULL );
                    } else if ( field.Type == DataField.BOOLEAN ) {
                        if ( TRUE.Equals( field.StringValue, StringComparison.OrdinalIgnoreCase ) ) {
                            writer.writeLiteral( TRUE );
                        } else {
                            writer.writeLiteral( FALSE );
                        }
                    } else if ( field.IsNumeric ) {
                        writer.writeNumber( field.StringValue );
                    } else if ( field.Type == DataField.FRAME ) {
                        writeFrame( (DataFrame)field.ObjectValue, writer );
                    } else {
                        object obj = field.ObjectValue;
                        if ( obj != null ) {
                            writer.writeString( obj.ToString() );
                        } else {
                            writer.writeLiteral( NULL );
                        }
                    }
                    if ( i + 1 < frame.Size ) {
                        writer.writeObjectSeparator();
                    }
                }

                if ( isArray )
                    writer.writeArrayClose();
                else
                    writer.writeObjectClose();

            } else {
                writer.writeObjectOpen();
                writer.writeObjectClose();
            }

            return;
        }

    }

}